From ceae5174c7036f54c210ad3a6fb2b7f920057ffb Mon Sep 17 00:00:00 2001
From: Yanteng Si <siyanteng@loongson.cn>
Date: Mon, 20 May 2024 10:20:21 +0800
Subject: [PATCH 115/360] LoongArch: fix some PCIE card not scanning properly

Fix some pcie card not scanning properly when
bus number is inconsistent during firmware and
kernel scan phases.

Signed-off-by: liuyun <liuyun@loongson.cn>
Signed-off-by: Tianli Xiong <xiongtianli@loongson.cn>
Signed-off-by: Yanteng Si <siyanteng@loongson.cn>
---
 drivers/pci/controller/pci-loongson.c | 34 +++++++++++++++++++++++++--
 1 file changed, 32 insertions(+), 2 deletions(-)

diff --git a/drivers/pci/controller/pci-loongson.c b/drivers/pci/controller/pci-loongson.c
index e2c6c1cfc054..d3e00c56797c 100644
--- a/drivers/pci/controller/pci-loongson.c
+++ b/drivers/pci/controller/pci-loongson.c
@@ -263,6 +263,36 @@ static void __iomem *pci_loongson_map_bus(struct pci_bus *bus,
 	return NULL;
 }
 
+static int pci_loongson_config_read(struct pci_bus *bus, unsigned int devfn,
+			    int where, int size, u32 *val)
+{
+	void __iomem *addr;
+
+	addr = bus->ops->map_bus(bus, devfn, where);
+	if (!addr) {
+		*val = ~0;
+		return PCIBIOS_DEVICE_NOT_FOUND;
+	}
+
+	if (size == 1)
+		*val = readb(addr);
+	else if (size == 2)
+		*val = readw(addr);
+	else
+		*val = readl(addr);
+	/*
+	 * fix some pcie card not scanning properly when bus number is
+	 * inconsistent during firmware and kernel scan phases.
+	 */
+	if (*val == 0x0 && where == PCI_VENDOR_ID) {
+		writel(*val, addr);
+		*val = readl(addr);
+	}
+
+
+	return PCIBIOS_SUCCESSFUL;
+}
+
 #ifdef CONFIG_OF
 
 static int loongson_map_irq(const struct pci_dev *dev, u8 slot, u8 pin)
@@ -286,7 +316,7 @@ static int loongson_map_irq(const struct pci_dev *dev, u8 slot, u8 pin)
 /* LS2K/LS7A accept 8/16/32-bit PCI config operations */
 static struct pci_ops loongson_pci_ops = {
 	.map_bus = pci_loongson_map_bus,
-	.read	= pci_generic_config_read,
+	.read	= pci_loongson_config_read,
 	.write	= pci_generic_config_write,
 };
 
@@ -413,7 +443,7 @@ const struct pci_ecam_ops loongson_pci_ecam_ops = {
 	.init	   = loongson_pci_ecam_init,
 	.pci_ops   = {
 		.map_bus = pci_loongson_map_bus,
-		.read	 = pci_generic_config_read,
+		.read	 = pci_loongson_config_read,
 		.write	 = pci_generic_config_write,
 	}
 };
-- 
2.20.1

